common/src/main/java/com/microsoft/identity/common/internal/util/PackageUtils.java

```diff
@@ -202,4 +208,41 @@ public static final X509Certificate getSelfSignedCert(final List<X509Certificate
 
         return selfSignedCert;
     }
+
+    /**
+     * A generalized API to perform signature validations on mentioned package.
+     *
+     * @param packageName   a {link List} of certificates to examine.
+     * @param context   the android context to use to search.
+     * @param validPackageSignatures    the valid package signatures to cross check
+     * @return a valid hash, if it is found.
+     * @throws ClientException
+     */
+    public static String signatureVerificationAndThrow(final String packageName, final Context context,
+                                                       final Iterator<String> validPackageSignatures)
+            throws ClientException {
+        try {
+            // Read all the certificates associated with the package name. In higher version of
+            // android sdk(>API 28, https://developer.android.com/reference/android/content/pm/PackageInfo#signatures),
+            // package manager will only returned the cert that is used to sign the
+            // APK. Even a cert is claimed to be issued by another certificates, sdk will return
+            // the signing cert. However, for the lower version of android, it will return all the
+            // certs in the chain. We need to verify that the cert chain is correctly chained up.
+            final List<X509Certificate> certs = readCertDataForApp(packageName, context);
+
+            final String signatureHash = PackageUtils.verifySignatureHash(certs, validPackageSignatures);
+
+            if (certs.size() > 1) {
+                PackageUtils.verifyCertificateChain(certs);
+            }
+
+            return signatureHash;
+        } catch (PackageManager.NameNotFoundException e) {
+            throw new ClientException(ErrorStrings.APP_PACKAGE_NAME_NOT_FOUND, e.getMessage(), e);
+        } catch (NoSuchAlgorithmException e) {
+            throw new ClientException(ErrorStrings.NO_SUCH_ALGORITHM, e.getMessage(), e);
+        } catch (final IOException | GeneralSecurityException e) {
+            throw new ClientException(ErrorStrings.BROKER_VERIFICATION_FAILED, e.getMessage(), e);
+        }
+    }
```

Comment:
This looks pretty much a copy / paste from the logic in BrokerValidator.java....let's avoid duplicating code. I think this logic can just sit in one place though that may require a bit of refactoring

Reply:
c9898c3